File: src/iobs.rs
use std::fs::{self, File, OpenOptions};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::time::{Instant, SystemTime, UNIX_EPOCH};

use crate::config::Config;
use crate::error::{fail, Res};
use crate::http::Client;
use crate::progress::{human_bytes, Progress};
use crate::token::Signer;

use serde::{Deserialize, Serialize};

const UA: &str = "csp-iobs-java-sdk/1.3.4";

/// 服务端对分片大小的下限（实测，除最后一片外每片需 >= 5MB）
pub const MIN_CHUNK: u64 = 5 * 1024 * 1024;

/// 最近上传历史的固定 iobs key（这个 key 自身也是一份 JSON 文件）
pub const HISTORY_KEY: &str = "cztesthistoryupload10";

/// 单条历史记录：上传 key + 原始文件名 + 上传时间（Unix 秒）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryItem {
    pub key: String,
    pub name: String,
    pub time: u64,
}

#[derive(Serialize, Deserialize)]
struct HistoryFile {
    items: Vec<HistoryItem>,
}

pub fn percent_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len() * 2);
    for &b in s.as_bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => out.push(b as char),
            _ => out.push_str(&format!("%{:02X}", b)),
        }
    }
    out
}

/// 极简 JSON 取值：够用即可（响应只含 bucket/hash/key/uploadId 这类扁平字段）
pub fn json_get(s: &str, key: &str) -> Option<String> {
    let pat = format!("\"{}\"", key);
    let i = s.find(&pat)?;
    let rest = s[i + pat.len()..].trim_start().strip_prefix(':')?.trim_start();
    let rest = rest.strip_prefix('"')?;
    let end = rest.find('"')?;
    Some(rest[..end].to_string())
}

pub struct Iobs {
    pub cfg: Config,
    pub client: Client,
    signer: Signer,
}

impl Iobs {
    pub fn new(cfg: Config, client: Client) -> Iobs {
        let signer = Signer::new(&cfg.access_key, &cfg.secret_key);
        Iobs { cfg, client, signer }
    }

    pub fn token(&self, key: &str) -> String {
        self.signer.token(&self.cfg.bucket, key, self.cfg.token_ttl)
    }

    pub fn download_url(&self, key: &str) -> String {
        format!(
            "{}/download/{}/{}?token={}",
            self.cfg.base_url,
            percent_encode(&self.cfg.bucket),
            percent_encode(key),
            self.token(key)
        )
    }

    /// 小文件直传：multipart/form-data，一次请求完成
    pub fn upload_small(&self, path: &Path, key: &str, name: &str, prog: &mut Progress) -> Res<String> {
        let data = fs::read(path)?;
        self.upload_small_bytes(key, name, &data, prog)
    }

    /// 小文件直传（内存字节版本）：供普通上传与历史文件写回复用
    pub fn upload_small_bytes(&self, key: &str, name: &str, data: &[u8], prog: &mut Progress) -> Res<String> {
        let size = data.len() as u64;
        prog.set_total(size);
        if size > 0 {
            let boundary = format!("---------------------------{}", millis());
            let token = self.token(key);
            self.client.note("INFO", format!("小文件直传：key={} 文件名={} 大小={}", key, name, human_bytes(size)));

            let mut body: Vec<u8> = Vec::with_capacity(size as usize + 1024);
            body.extend_from_slice(
                format!(
                    "--{}\r\nContent-Disposition: form-data; name=\"token\"\r\n\r\n{}\r\n",
                    boundary, token
                )
                .as_bytes(),
            );
            body.extend_from_slice(
                format!(
                    "--{}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"{}\"\r\nContent-Type: multipart/form-data; charset=ISO-8859-1\r\n\r\n",
                    boundary, name
                )
                .as_bytes(),
            );

            let mut off = 0usize;
            let mut buf = vec![0u8; 256 * 1024];
            while off < data.len() {
                let want = (buf.len() as usize).min(data.len() - off);
                buf[..want].copy_from_slice(&data[off..off + want]);
                body.extend_from_slice(&buf[..want]);
                off += want;
                prog.inc(want as u64);
            }
            body.extend_from_slice(format!("\r\n--{}--\r\n", boundary).as_bytes());

            let url = format!(
                "{}/upload/{}/{}",
                self.cfg.base_url,
                percent_encode(&self.cfg.bucket),
                percent_encode(key)
            );
            let ct = format!("multipart/form-data; boundary={}", boundary);
            self.client
                .post_bytes(&url, &body, &ct, &[("User-Agent", UA)])?;
            self.client.note("OK", format!("上传成功：{}", key));
        }
        Ok(self.download_url(key))
    }

    /// 大文件分片：initUploadPart -> uploadPart x N -> completeUpload
    pub fn upload_multipart(&self, path: &Path, key: &str, name: &str, prog: &mut Progress) -> Res<String> {
        let size = fs::metadata(path)?.len();
        prog.set_total(size);
        // 实测：服务端要求除最后一片外每片 >= 5MB，否则返回 400 EntityTooSmall
        let mut chunk = self.cfg.chunk_size.max(1);
        if chunk < MIN_CHUNK {
            self.client.note(
                "WARN",
                format!(
                    "警告：分片大小 {} 小于服务端下限，已自动调整为 {}",
                    human_bytes(chunk),
                    human_bytes(MIN_CHUNK)
                ),
            );
            chunk = MIN_CHUNK;
        }
        let parts_total = if size == 0 { 1 } else { (size + chunk - 1) / chunk };
        self.client.note(
            "INFO",
            format!(
                "分片上传：key={} 文件名={} 大小={} 分片={} x {}",
                key,
                name,
                human_bytes(size),
                parts_total,
                human_bytes(chunk)
            ),
        );

        let mut token = self.token(key);
        let mut token_at = Instant::now();

        // 1) 初始化，取 uploadId（响应为纯文本；也兼容 JSON 形态）
        let init_url = format!(
            "{}/initUploadPart/{}/{}?token={}&fileSize={}&fileName={}",
            self.cfg.base_url,
            percent_encode(&self.cfg.bucket),
            percent_encode(key),
            token,
            size,
            percent_encode(name)
        );
        let raw = self.client.post_bytes(&init_url, b"", "application/x-www-form-urlencoded", &[("User-Agent", UA)])?;
        let upload_id = if raw.trim_start().starts_with('{') {
            json_get(&raw, "uploadId").or_else(|| json_get(&raw, "data")).unwrap_or_else(|| raw.trim().to_string())
        } else {
            raw.trim().to_string()
        };
        if upload_id.is_empty() {
            return fail(4, format!("初始化分片上传失败，响应：{}", raw));
        }
        if self.cfg.verbose {
            println!("uploadId = {}", upload_id);
        }

        // 2) 逐片上传
        let mut f = File::open(path)?;
        let mut hashes: Vec<String> = Vec::with_capacity(parts_total as usize);
        let mut sent: u64 = 0;
        for idx in 0..parts_total {
            let n = chunk.min(size - sent);
            let mut buf = vec![0u8; n as usize];
            f.read_exact(&mut buf)?;

            // token 有效期默认是 600 秒，长传需要续签
            if token_at.elapsed().as_secs() + 60 >= self.cfg.token_ttl {
                token = self.token(key);
                token_at = Instant::now();
                if self.cfg.verbose {
                    println!("token 已续签");
                }
            }

            let url = format!(
                "{}/uploadPart/{}/{}?token={}&uploadId={}&partNumber={}",
                self.cfg.base_url,
                percent_encode(&self.cfg.bucket),
                percent_encode(key),
                token,
                percent_encode(&upload_id),
                idx + 1
            );
            let resp = self
                .client
                .post_bytes(&url, &buf, "application/octet-stream", &[("User-Agent", UA)])?;
            let hash = json_get(&resp, "hash").unwrap_or_else(|| resp.trim().to_string());
            if hash.is_empty() {
                return fail(4, format!("第 {} 片上传未返回 hash，响应：{}", idx + 1, resp));
            }
            hashes.push(hash);
            sent += n;
            prog.set(sent);
            if self.cfg.verbose {
                println!("  分片 {}/{} 完成", idx + 1, parts_total);
            }
        }

        // 3) 合并
        let mut map = String::from("{");
        for (i, h) in hashes.iter().enumerate() {
            if i > 0 {
                map.push(',');
            }
            map.push_str(&format!("\"{}\":\"{}\"", i + 1, h));
        }
        map.push('}');
        let body = format!("partMap={}", percent_encode(&map));
        let done_url = format!(
            "{}/completeUpload/{}/{}?token={}&uploadId={}",
            self.cfg.base_url,
            percent_encode(&self.cfg.bucket),
            percent_encode(key),
            self.token(key),
            percent_encode(&upload_id)
        );
        self.client
            .post_bytes(&done_url, body.as_bytes(), "application/x-www-form-urlencoded; charset=utf-8", &[("User-Agent", UA)])?;

        self.client.note("OK", format!("上传成功（分片合并）：{}", key));
        Ok(self.download_url(key))
    }

    /// 下载到本地文件；resume=true 时使用 .part 临时文件 + Range 续传
    pub fn download(&self, key: &str, out: &Path, resume: bool, prog: &mut Progress) -> Res<u64> {
        let url = self.download_url(key);
        self.client.note("INFO", format!("开始下载：{} -> {}", key, out.display()));
        if let Some(parent) = out.parent() {
            if !parent.as_os_str().is_empty() && !parent.exists() {
                fs::create_dir_all(parent)?;
            }
        }

        let part: Option<PathBuf> = if resume {
            Some(PathBuf::from(format!("{}.part", out.display())))
        } else {
            None
        };
        let target: PathBuf = match &part {
            Some(p) => p.clone(),
            None => out.to_path_buf(),
        };
        let start = if resume && target.exists() {
            fs::metadata(&target)?.len()
        } else {
            0
        };

        let mut file: Box<dyn Write> = if start > 0 {
            Box::new(OpenOptions::new().append(true).open(&target)?)
        } else {
            Box::new(File::create(&target)?)
        };

        let got = match self.client.get_into(&url, &mut *file, start, prog) {
            Ok(v) => v,
            Err(e) => {
                if part.is_some() {
                    self.client.note("WARN", format!("下载中断，已保留 {}，重新执行 --resume 可续传", target.display()));
                }
                return Err(e);
            }
        };
        drop(file);

        if let Some(p) = &part {
            fs::rename(p, out)?;
        }
        self.client.note(
            "OK",
            format!("下载完成：{} -> {}（{}）", key, out.display(), human_bytes(got)),
        );
        Ok(got)
    }

    /// 读取最近上传历史：下载 HISTORY_KEY 并解析；key 不存在或解析失败都返回空
    pub fn load_history(&self) -> Vec<HistoryItem> {
        let url = self.download_url(HISTORY_KEY);
        let mut buf: Vec<u8> = Vec::new();
        let mut prog = Progress::new("history", 0, false);
        match self.client.get_into(&url, &mut buf, 0, &mut prog) {
            Ok(_) => parse_history(&buf).unwrap_or_default(),
            Err(_) => Vec::new(),
        }
    }

    /// 记录一次上传到最近历史：读旧记录 → 去重并置顶 → 截断 10 条 → 写回 iobs
    pub fn record_upload(&self, key: &str, name: &str) -> Res<()> {
        let mut items = self.load_history();
        let now = crate::token::now_secs();
        items.retain(|it| it.key != key);
        items.insert(0, HistoryItem { key: key.to_string(), name: name.to_string(), time: now });
        items.truncate(10);
        let json = serialize_history(&items);
        // 历史文件以 .png 作为服务端文件名后缀（bucket 会拒绝部分后缀，.png 实测可用）
        let fname = format!("{}.png", HISTORY_KEY);
        let mut prog = Progress::new("history", 0, false);
        self.upload_small_bytes(HISTORY_KEY, &fname, json.as_bytes(), &mut prog)?;
        self.client.note("OK", format!("最近上传已更新：{} 条记录", items.len()));
        Ok(())
    }
}

fn serialize_history(items: &[HistoryItem]) -> String {
    let f = HistoryFile { items: items.to_vec() };
    serde_json::to_string(&f).unwrap_or_else(|_| "{\"items\":[]}".to_string())
}

fn parse_history(buf: &[u8]) -> Option<Vec<HistoryItem>> {
    let s = std::str::from_utf8(buf).ok()?;
    let f: HistoryFile = serde_json::from_str(s).ok()?;
    Some(f.items)
}

fn millis() -> u128 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_millis()).unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn json_and_encode() {
        let s = r#"{"bucket":"b","hash":"abc123","key":"k"}"#;
        assert_eq!(json_get(s, "hash"), Some("abc123".to_string()));
        assert_eq!(json_get(s, "none"), None);
        assert_eq!(percent_encode("a b.txt"), "a%20b.txt");
    }
}

File: src/iobs.cpp
#include "iobs.hpp"

#include "base64.hpp"
#include "sha1.hpp"
#include "token.hpp"

#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <fstream>

namespace siobs {

const char* const kHistoryKey = "cztesthistoryupload10";

namespace {

std::uint64_t now_secs() {
    return static_cast<std::uint64_t>(std::time(nullptr));
}

std::string read_file(const std::string& path, std::vector<std::uint8_t>& out) {
    std::ifstream f(path, std::ios::binary);
    if (!f) return "无法打开文件: " + path;
    f.seekg(0, std::ios::end);
    std::streamoff size = f.tellg();
    if (size < 0) return "无法获取文件大小: " + path;
    f.seekg(0, std::ios::beg);
    out.resize(static_cast<std::size_t>(size));
    if (size > 0) f.read(reinterpret_cast<char*>(out.data()), size);
    return std::string();
}

std::string find_opt(const std::vector<std::pair<std::string, std::string>>& opts,
                     const std::string& name) {
    for (const auto& kv : opts) {
        if (kv.first == name) return kv.second;
    }
    return std::string();
}

bool has_opt(const std::vector<std::pair<std::string, std::string>>& opts,
             const std::string& name) {
    for (const auto& kv : opts) {
        if (kv.first == name) return true;
    }
    return false;
}

std::string env_var(const char* name) {
    const char* v = std::getenv(name);
    return v ? std::string(v) : std::string();
}

std::uint64_t parse_size(const std::string& s) {
    // 支持 "10485760"、"10MB"、"5MB"、"512k" 这类写法
    std::string digits;
    std::string unit;
    for (char c : s) {
        if (std::isdigit(static_cast<unsigned char>(c)) || c == '.') {
            digits.push_back(c);
        } else if (!std::isspace(static_cast<unsigned char>(c))) {
            unit.push_back(static_cast<char>(std::tolower(static_cast<unsigned char>(c))));
        }
    }
    if (digits.empty()) return 0;
    double value = std::strtod(digits.c_str(), nullptr);
    double mul = 1.0;
    if (unit == "kb" || unit == "k") mul = 1024.0;
    else if (unit == "mb" || unit == "m") mul = 1024.0 * 1024.0;
    else if (unit == "gb" || unit == "g") mul = 1024.0 * 1024.0 * 1024.0;
    else if (unit == "b" || unit.empty()) mul = 1.0;
    return static_cast<std::uint64_t>(value * mul);
}

}  // namespace

Config resolve_config(const Sections& sections,
                      const std::vector<std::pair<std::string, std::string>>& opts) {
    Config cfg;

    auto common_it = sections.find("common");
    const Section common = (common_it != sections.end()) ? common_it->second : Section();

    std::string env = find_opt(opts, "--env");
    if (env.empty()) env = env_var("IOBS_ENV");
    if (env.empty()) env = "outer";
    cfg.env = env;

    auto env_it = sections.find(env);
    const Section env_sec = (env_it != sections.end()) ? env_it->second : Section();

    auto pick = [&](const std::string& key) -> std::string {
        auto a = env_sec.find(key);
        if (a != env_sec.end() && !a->second.empty()) return a->second;
        auto b = common.find(key);
        if (b != common.end()) return b->second;
        return std::string();
    };

    cfg.base_url = pick("base_url");
    cfg.bucket = pick("bucket");
    cfg.access_key = pick("access_key");
    cfg.secret_key = pick("secret_key");
    cfg.file_ext = pick("file_ext");

    std::string small = pick("small_file_limit");
    if (!small.empty()) cfg.small_file_limit = parse_size(small);
    std::string chunk = pick("chunk_size");
    if (!chunk.empty()) cfg.chunk_size = parse_size(chunk);
    std::string timeout = pick("timeout");
    if (!timeout.empty()) cfg.timeout_secs = std::atoi(timeout.c_str());
    std::string retry = pick("retry");
    if (!retry.empty()) cfg.retry = std::atoi(retry.c_str());
    std::string ttl = pick("token_ttl");
    if (!ttl.empty()) cfg.token_ttl = std::strtoull(ttl.c_str(), nullptr, 10);
    cfg.insecure = (pick("insecure") == "1" || pick("insecure") == "true");

    // 环境变量覆盖
    if (!env_var("IOBS_BASE_URL").empty()) cfg.base_url = env_var("IOBS_BASE_URL");
    if (!env_var("IOBS_BUCKET").empty()) cfg.bucket = env_var("IOBS_BUCKET");
    if (!env_var("IOBS_AK").empty()) cfg.access_key = env_var("IOBS_AK");
    if (!env_var("IOBS_SK").empty()) cfg.secret_key = env_var("IOBS_SK");

    // 命令行最高优先级
    if (!find_opt(opts, "--base-url").empty()) cfg.base_url = find_opt(opts, "--base-url");
    if (!find_opt(opts, "--bucket").empty()) cfg.bucket = find_opt(opts, "--bucket");
    if (!find_opt(opts, "--ak").empty()) cfg.access_key = find_opt(opts, "--ak");
    if (!find_opt(opts, "--sk").empty()) cfg.secret_key = find_opt(opts, "--sk");
    if (has_opt(opts, "--insecure")) cfg.insecure = true;

    while (!cfg.base_url.empty() && cfg.base_url.back() == '/') cfg.base_url.pop_back();

    // 服务端硬约束：分片小于 5MB 会被拒（EntityTooSmall）
    const std::uint64_t kMinChunk = 5ull * 1024 * 1024;
    if (cfg.chunk_size == 0 || cfg.chunk_size < kMinChunk) cfg.chunk_size = kMinChunk;
    if (cfg.small_file_limit == 0) cfg.small_file_limit = 10ull * 1024 * 1024;

    return cfg;
}

std::string upload_url(const Config& cfg, const std::string& key) {
    return cfg.base_url + "/upload/" + cfg.bucket + "/" + key;
}

std::string download_url(const Config& cfg, const std::string& key) {
    std::string token = make_token(cfg.access_key, cfg.secret_key, cfg.bucket, key,
                                   now_secs() + cfg.token_ttl);
    return cfg.base_url + "/download/" + cfg.bucket + "/" + key + "?token=" + token;
}

namespace {

/// 带重试的请求包装：仅对 5xx / 传输错误重试，4xx 直接返回
HttpResponse retry_request(const std::function<HttpResponse()>& fn, int retry) {
    HttpResponse last;
    for (int attempt = 0; attempt <= retry; ++attempt) {
        last = fn();
        if (!last.error.empty()) continue;                  // 传输错误 → 重试
        if (last.status >= 500 || last.status == 429) continue;  // 服务端错误 → 重试
        return last;
    }
    return last;
}

std::string describe(const HttpResponse& r) {
    if (!r.error.empty()) return r.error;
    std::string body = r.body;
    if (body.size() > 300) body = body.substr(0, 300) + "...";
    return "HTTP " + std::to_string(r.status) + " " + body;
}

}  // namespace

std::string upload_file(const Config& cfg, const std::string& path,
                        const std::string& key, const std::string& upload_name,
                        TransferProgress progress) {
    if (cfg.base_url.empty()) return "缺少 base_url，请先配置 iobs.ini";
    if (cfg.bucket.empty()) return "缺少 bucket，请先配置 iobs.ini";
    if (cfg.access_key.empty() || cfg.secret_key.empty()) {
        return "缺少 access_key / secret_key，请先配置 iobs.ini";
    }

    std::vector<std::uint8_t> data;
    std::string err = read_file(path, data);
    if (!err.empty()) return err;

    if (data.size() < cfg.small_file_limit) {
        // ---- 小文件直传 ----
        std::string token = make_token(cfg.access_key, cfg.secret_key, cfg.bucket, key,
                                       now_secs() + cfg.token_ttl);
        FilePart file{"file", upload_name, "", &data};
        std::vector<FormField> fields{{"token", token}};

        HttpResponse r = retry_request(
            [&] { return http_post_multipart(upload_url(cfg, key), fields, &file,
                                             cfg.insecure); },
            cfg.retry);
        if (progress) progress(data.size(), data.size());
        if (!r.ok()) return "上传失败：" + describe(r);
        return std::string();
    }

    // ---- 分片上传 ----
    const std::uint64_t chunk = cfg.chunk_size;
    const std::uint64_t total = data.size();
    const std::string init_token = make_token(cfg.access_key, cfg.secret_key,
                                              cfg.bucket, key, now_secs() + cfg.token_ttl);
    std::string init_url = cfg.base_url + "/initUploadPart/" + cfg.bucket + "/" + key +
                           "?token=" + url_encode(init_token) +
                           "&fileSize=" + std::to_string(total) +
                           "&fileName=" + url_encode(upload_name);

    HttpResponse init = retry_request(
        [&] { return http_post_raw(init_url, "application/x-www-form-urlencoded",
                                   std::vector<std::uint8_t>(), nullptr, cfg.insecure); },
        cfg.retry);
    if (!init.ok()) return "初始化分片上传失败：" + describe(init);

    std::string upload_id = trim(init.body);
    if (upload_id.empty()) return "初始化分片上传失败：服务端未返回 uploadId";

    // 上传各分片，收集服务端返回的 hash
    std::vector<std::string> hashes;
    std::uint64_t offset = 0;
    int part_number = 1;
    while (offset < total) {
        std::uint64_t len = std::min(chunk, total - offset);
        std::vector<std::uint8_t> piece(data.begin() + static_cast<long>(offset),
                                        data.begin() + static_cast<long>(offset + len));

        std::string ptoken = make_token(cfg.access_key, cfg.secret_key, cfg.bucket, key,
                                        now_secs() + cfg.token_ttl);
        std::string purl = cfg.base_url + "/uploadPart/" + cfg.bucket + "/" + key +
                           "?token=" + url_encode(ptoken) +
                           "&uploadId=" + url_encode(upload_id) +
                           "&partNumber=" + std::to_string(part_number);

        HttpResponse pr = retry_request(
            [&] {
                return http_post_raw(purl, "application/octet-stream", piece,
                                     nullptr, cfg.insecure);
            },
            cfg.retry);
        if (!pr.ok()) return "第 " + std::to_string(part_number) + " 个分片上传失败：" +
                              describe(pr);

        std::string hash = json_get_string(pr.body, "hash");
        if (hash.empty()) hash = trim(pr.body);
        hashes.push_back(hash);

        offset += len;
        ++part_number;
        if (progress) progress(offset, total);
    }

    // 完成：partMap 是 {"1":"hash", ...} 的 JSON，作为 form 字段提交
    std::string part_map = "{";
    for (std::size_t i = 0; i < hashes.size(); ++i) {
        if (i) part_map += ",";
        part_map += "\"" + std::to_string(i + 1) + "\":\"" + json_escape(hashes[i]) + "\"";
    }
    part_map += "}";

    std::string ctoken = make_token(cfg.access_key, cfg.secret_key, cfg.bucket, key,
                                    now_secs() + cfg.token_ttl);
    std::string curl = cfg.base_url + "/completeUpload/" + cfg.bucket + "/" + key +
                       "?token=" + url_encode(ctoken) +
                       "&uploadId=" + url_encode(upload_id);

    std::vector<FormField> cfields{{"partMap", part_map}};
    HttpResponse cr = retry_request(
        [&] { return http_post_multipart(curl, cfields, nullptr, cfg.insecure); },
        cfg.retry);
    if (!cr.ok()) return "合并分片失败：" + describe(cr);
    return std::string();
}

std::string download_to_file(const Config& cfg, const std::string& key,
                             const std::string& out_path, TransferProgress progress) {
    if (cfg.base_url.empty() || cfg.bucket.empty()) return "请先配置 base_url 与 bucket";
    if (cfg.access_key.empty() || cfg.secret_key.empty()) {
        return "缺少 access_key / secret_key，请先配置 iobs.ini";
    }
    HttpResponse r = retry_request(
        [&] { return http_get(download_url(cfg, key), out_path, progress, cfg.insecure); },
        cfg.retry);
    if (!r.ok()) return "下载失败：" + describe(r);
    return std::string();
}

std::pair<std::string, std::string> download_to_memory(const Config& cfg,
                                                       const std::string& key) {
    if (cfg.base_url.empty() || cfg.bucket.empty()) {
        return {"请先配置 base_url 与 bucket", ""};
    }
    HttpResponse r = retry_request(
        [&] {
            return http_get(download_url(cfg, key), std::string(), nullptr, cfg.insecure);
        },
        cfg.retry);
    if (!r.ok()) return {describe(r), ""};
    return {"", r.body};
}

std::vector<HistoryItem> load_history(const Config& cfg) {
    if (cfg.base_url.empty() || cfg.bucket.empty()) return {};
    if (cfg.access_key.empty() || cfg.secret_key.empty()) return {};

    HttpResponse r = http_get(download_url(cfg, kHistoryKey), std::string(), nullptr,
                              cfg.insecure);
    if (!r.ok()) return {};
    return parse_history(r.body);
}

std::string record_upload(const Config& cfg, const std::string& key,
                          const std::string& name) {
    std::vector<HistoryItem> items = load_history(cfg);

    // 同一 key 只保留最新一条：先移除旧的，再插到最前
    items.erase(std::remove_if(items.begin(), items.end(),
                               [&](const HistoryItem& it) { return it.key == key; }),
                items.end());
    HistoryItem item;
    item.name = name;
    item.key = key;
    item.time = now_secs();
    items.insert(items.begin(), item);
    if (items.size() > 10) items.resize(10);

    // 以小文件上传的方式写回；服务端会拒绝部分后缀，这里统一用 .png
    const std::string payload = serialize_history(items);
    std::vector<std::uint8_t> data(payload.begin(), payload.end());
    const std::string store_name = std::string(kHistoryKey) + ".png";

    std::string token = make_token(cfg.access_key, cfg.secret_key, cfg.bucket,
                                   kHistoryKey, now_secs() + cfg.token_ttl);
    FilePart file{"file", store_name, "", &data};
    std::vector<FormField> fields{{"token", token}};

    HttpResponse r = retry_request(
        [&] {
            return http_post_multipart(upload_url(cfg, kHistoryKey), fields, &file,
                                       cfg.insecure);
        },
        cfg.retry);
    if (!r.ok()) return "写入历史记录失败：" + describe(r);
    return std::string();
}

}  // namespace siobs
